Descriptor checksums

sipa opened the pull request:
This adds support for a descriptor-specific 8-character checksum.

Descriptors may optionally be suffixed with a `#` plus these 8 checksum characters. Any descriptor that contains a `#` at the end must be followed by a valid checksum. If the `#` is missing entirely, it is valid without checksum.

All RPCs are updated to report descriptors that include the checksum. On input, they are optional except in `deriveaddress` and `importmulti`, which require descriptors which include a checksum.

A new RPC is also added to analyse descriptors (`getdescriptorinfo`), which can be used to compute the checksum for a descriptor without.

sipa committed:
Descriptor checksum

sipa committed:
Add getdescriptorinfo to compute checksum

instagibbs commented:
Can a motivation for the placement be given? Accidentally eliding it for whatever reason neuters the protection while still maintaining a valid descriptor, but clearly it seems simpler from an implementation and compatibility perspective.

sipa commented:
@gsanders Well for critical RPCs the plan is that the checksum won't be optional. I just haven't included that in this PR as it means adapting a bunch of tests, which I only want to do once the checksum algorithm is final.

MeshCollider commented:
Concept ACK

promag commented:
Concept ACK.

It'd be nice to read a draft to update doc/descriptors.md.

sipa committed:
Add checksums to descriptors.md

sipa commented:
Added a section to `doc/descriptors.md`.

promag commented on doc/descriptors.md:
```diff
@@ -169,3 +169,20 @@ existing Bitcoin Core wallets, a convenience function `combo` is
+restricted to substituting characters in `0123456789()[],'/*abcdefgh@:$%{}`
+for others in that set and changes in letter case, up to 4 errors will always
+be detected in descriptors up to 501 characters, and up to 3 errors in longer
+ones. For larger numbers of errors, or other types of errors, the is a
```
s/the is/there is?

sipa replied:
Fixed.

practicalswift commented on src/test/descriptor_tests.cpp:
```diff
@@ -28,13 +28,27 @@ constexpr int HARDENED = 2; // Derivation needs access to private keys
+        if (a_check) a = a.substr(0, a.size() - 9);
+        if (b_check) b = b.substr(0, b.size() - 9);
+    }
+    if (a != b) fprintf(stderr, "%s != %s\n", a.c_str(), b.c_str());
```
Should this a debug print be left in here?

sipa replied:
Fixed.

practicalswift commented on src/script/descriptor.cpp:
```diff
@@ -19,6 +19,97 @@
+////////////////////////////////////////////////////////////////////////////
+
+// This section implements a checksum algorithm for descriptors with the following properties:
+// * Every 1 character substitition error counts as 1 or 2 symbol errors, but:
```
Subsitute this substitition with substitution.

sipa replied:
Fixed.

Sjors commented on src/rpc/misc.cpp:
```diff
@@ -143,6 +143,44 @@ static UniValue createmultisig(const JSONRPCRequest& request)
+            "  \"descriptor\" : \"desc\",       (string) The descriptor in canonical form, without private keys\n"
+            "  \"isrange\" : true|false,        (boolean) Whether the descriptor is ranged\n"
+            "  \"issolvable\" : true|false,     (boolean) Whether the descriptor is solvable\n"
+            "  \"isprivate\" : true|false,      (boolean) Whether the input descriptor contained at least one private key\n"
```
Nit: rename `isprivate` to `hasprivatekey` (or `contains...`).

sipa replied:
Done.

Sjors reviewed (commented):
Concept ACK, will review shortly. Agree that `deriveaddress` and `importmulti` should require a checksum. The introduction of `getdescriptorinfo` means there's no need to make that opt-out.

Do I understand correctly that the checksum is either based on the canonical form, i.e. based on public keys, or skips keys altogether (but that seems suboptimal when these keys are not in checksummed xpub form)? Otherwise the result of  `getdescriptorinfo` could be confusing if you feed it an `xpriv`.

Maybe return a warning if a user does provide an `xpriv` that they should clear their shell command history (and generally recommend either not doing that, or providing a safer method like #15346).

MeshCollider commented:
The surrounding code looks good other than the comments above, haven't reviewed the actual checksum code itself yet

laanwj commented on src/rpc/misc.cpp:
```diff
@@ -143,6 +143,44 @@ static UniValue createmultisig(const JSONRPCRequest& request)
+            "}\n"
+            },
+            RPCExamples{
+                "Analyse a descriptor\n" +
```
might want to mention that this example only works on mainnet

sipa replied:
Fixed by changing to a pubkey-only example.

laanwj commented:
utACK, code changes look good to me, haven't checked any of the magic numbers in `PolyMod`

sipa committed:
Make descriptor checksums mandatory in deriveaddresses and importmulti

sipa commented:
Several changes:
* Addressed all comments
* Finalized the checksum design (and switched to a slightly better generator)
* Added explanation (incl. Sage code) of the checksum
* Made checksums mandatory in `deriveaddresses` and `importmulti`
* Expanded and updated tests

sipa commented on src/script/descriptor.cpp:
```diff
@@ -19,6 +19,117 @@
+ * - Selecting just the set of generators that guarantee detecting 4 errors in a window of length 512.
+ * - Selecting one of those with best worst-case behavior for 5 errors in windows of length up to 512.
+ *
+ * The generator and the constants to implement it can be verified using this Sage code:
```
@gmaxwell @sdaftuar What do you think about the inclusion of the Sage code here? I could do the same for the Bech32.

sipa replied:
I think that's overkill.

Sjors commented:
Breaks Travis due to #14918.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -19,6 +19,117 @@
+
+// This section implements a checksum algorithm for descriptors with the following properties:
+// * Every 1 character substitution error counts as 1 or 2 symbol errors, but:
+//   * An error substituting a character from 0123456789()[],'/*abcdefgh@:$%{} for another in
```
Maybe flip this around, and explain a bit more how symbol error count works:

```c++
// * Mistakes in a descriptor string are measured in symbol errors. A higher symbol
//   error count is more difficult to detect, because it becomes indistinguishable
//   from a different descriptor.
// * An error substituting a character from 0123456789()[],'/*abcdefgh@:$%{} for
//   another in that set always counts as 1 symbol error.
// * A case error always counts as 1 symbol error.
// * Any other 1 character substitution error counts as 1 or 2 symbol errors.
// * Note that hex encoded keys are covered by these special characters, whereas xprivs
//   and xpubs use different characters, but already have their own checksum mechanism.
//   Function names like "multi()" use different characters, but mistakes would generally
//   result in an unparseable descriptor. 
```

Some of this is also explained in `DescriptorChecksum` in different wording.

sipa replied:
Cool, that's more clear. I've included it with some copy-editing.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -19,6 +19,117 @@
+    static std::string INPUT_CHARSET =
+        "0123456789()[],'/*abcdefgh@:$%{}"
+        "IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~"
+        "ijklmnopqrstuvwxyzABCDEFGH`#\"\\ ";
```
Nit: putting `ABCDEFGH` and `abcdefgh` all the way to the right would make it a bit more clear that they intentionally have the same offset (for some reason Github doesn't believe in fixed-width font for code, but even in editors it would be more clear).

sipa replied:
What platform are you using? It's fixed width here, and I prefer to keep them in alphabetical order.

sipa replied:
Github renders in a fixed-width font here. I don't think this concern weighs up against keeping the characters in alphabetical ordering (to the extent possible).

practicalswift replied:
FWIW, `abc...`  and `ABC...` are perfectly aligned here :-)

A non-fixed width GitHub code view is surely not intentional -- I suggest reporting to GitHub! :-)

Sjors replied:
Nvm, it was just the escape characters in the bottom line that made it look misaligned. That or I was A/B-tested.

Sjors commented on src/rpc/misc.cpp:
```diff
@@ -193,7 +233,7 @@ UniValue deriveaddresses(const JSONRPCRequest& request)
-    auto desc = Parse(desc_str, provider);
+    auto desc = Parse(desc_str, provider, true);
     if (!desc) {
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Invalid descriptor"));
```
It would be nice to have a dinstinct error for a missing checksum.

instagibbs replied:
In general the errors are non-existent. Like, a space causes it to be rejected. Maybe a followup PR to make some common error cases printed?

sipa replied:
Agree, but I'd prefer to do that in a different change.

Sjors replied:
I think that's fine because both descriptor enhance `importmulti` and `deriveaddress` are new anyway. We can make them more user friendly later.

Sjors reviewed (commented):
tACK 4f95087 modulo RPC help syntax

I can't vouch for the actual math, but the documentation, Sage code, tests and Python re-implementation are comforting. One way, perhaps overkill, to sanity check that the checksum works as intended is to generate a whole bunch of deterministic typos and see that they are indeed detected.

sipa commented:
Rebased.

gmaxwell commented:
ACK

instagibbs commented on src/rpc/misc.cpp:
```diff
@@ -143,6 +143,46 @@ static UniValue createmultisig(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() != 1) {
+        throw std::runtime_error(
+            RPCHelpMan{"getdescriptorinfo",
+            {"\nAnalyses a descriptor.\n"},
```
Analyze?

flack replied:
@instagibbs it's funny, I stumbled across this, too, and did a little search on analyse/analyze in the code base, it's almost evenly split. So it would probably be better to do a separate PR that standardizes (standardises?) on either the British or American spelling

gmaxwell replied:
I am not a fan of PRs to go around switching between different english styles, especially in comments.  All of them are valid, all can be mutually read by english speakers.  Trying to maintain consistency would just mean a never ending sequence of fixups PRs.   

We have a finite amount of resources to handle changes, they should be conserved for efforts that improve the capability or reliability of the software.

flack replied:
I'm not saying it should necessarily be changed, but changing it in this PR only will not really add anything wrt consistency either (since analyze/analyse appear in more or less the same frequency in the code base). So... leave it as is?

instagibbs replied:
I had no idea it was a valid word, haha. I retract the comment!

instagibbs replied:
I thought it was a misspelling of "analysis" which made no grammatical sense.

instagibbs reviewed (changes requested):
Please add a single functional test for deriveaddresses and importmulti lacking the checksum, I compiled out the mandatory flag for them and tests seem to pass.

sipa commented:
Rebased and added a test to `deriveaddresses` and `importmulti` to test for missing checksum.

instagibbs reviewed (approved):
tests now fail when I allow no checksum

tACK https://github.com/bitcoin/bitcoin/pull/15368/commits/34164546d8f9c2c1f0b05e2b2f8b83a3ac16eec3

fanquake commented:
Can the `deriveaddresses` RPC example be updated with a checksum, otherwise it will no-longer work.

Looks like the descriptor should be (from `getdescriptorinfo`):
```
"wpkh([d34db33f/84'/0'/0']xpub6DJ2dNUysrn5Vt36jH2KLBT2i1auw1tTSSomg8PhqNiUtx8QX2SvC9nrHu81fT41fvDUnhMjEzQgXnQjKEu3oaqMSzhSrHMxyyoEAmUHQbY/0/*)#trd0mf0l"
```

promag commented on src/script/descriptor.h:
```diff
@@ -63,7 +63,7 @@ struct Descriptor {
 
 /** Parse a descriptor string. Included private keys are put in out. Returns nullptr if parsing fails. */
-std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out);
+std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out, bool require_checksum = false);
```
Instead of adding the new argument with a default value, why not create a new function `ParseChecked`? From the call site it is more clear, especially since it's statically defined where checksum is needed.

promag replied:
Or you could do `auto desc = Parse(desc_str, provider, /* require_checksum = */ true);`.

sipa replied:
Did the latter.

promag reviewed (commented):
Needs release for the new RPC. `deriveaddress` release notes already points to descriptors.md.

sipa commented:
@promag Needs release?

sipa commented on test/functional/rpc_deriveaddresses.py:
```diff
@@ -14,36 +15,37 @@ def set_test_params(self):
 
-        descriptor_pubkey = "wpkh(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/0)"
-        address = "bcrt1qjqmxmkpmxt80xz4y3746zgt0q3u3ferr34acd5"
+        descriptor = "wpkh(tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK/1/1/0)"
```
Done.

sipa commented:
@promag I've just added a "needs release notes" label for now, as it intersects with the notes added for `deriveaddresses` and `importmulti`.

promag reviewed (commented):
Could have an "empty checksum" test. Feel free to ignore my nits. LGTM code wise.

sipa commented on src/script/descriptor.h:
```diff
@@ -63,7 +63,7 @@ struct Descriptor {
 };
 
 /** Parse a descriptor string. Included private keys are put in out. Returns nullptr if parsing fails. */
```
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -20,6 +20,125 @@
+            clscount = 0;
+        }
+    }
+    if (clscount) c = PolyMod(c, cls);
```
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -20,6 +20,125 @@
+    for (int j = 0; j < 8; ++j) c = PolyMod(c, 0); // Shift further to determine the checksum.
+    c ^= 1; // Prevent appending zeroes from not affecting the checksum.
+
+    std::string ret;
```
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -20,6 +20,125 @@
+    int cls = 0;
+    int clscount = 0;
+    for (auto ch : span) {
+        auto pos = INPUT_CHARSET.find(ch);
```
I don't think the extra code is worth the performance gain.

sipa commented on src/rpc/misc.cpp:
```diff
@@ -143,6 +143,46 @@ static UniValue createmultisig(const JSONRPCRequest& request)
+    FlatSigningProvider provider;
+    auto desc = Parse(request.params[0].get_str(), provider);
+    if (!desc) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Invalid descriptor"));
```
There is (rpc_deriveaddresses functional test).

promag replied:
Not for getdescriptorinfo.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -751,11 +875,25 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
+    if (check_split.size() == 2) {
+        auto checksum = DescriptorChecksum(check_split[0]);
+        if (checksum.size() == 0) return nullptr; // Invalid characters in payload
+        if ((size_t)check_split[1].size() != checksum.size()) return nullptr; // Unexpected length for checksum
```
Good idea, done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -751,11 +875,25 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
+    if (check_split.size() == 1 && require_checksum) return nullptr; // Missing checksum
+    if (check_split.size() == 2) {
+        auto checksum = DescriptorChecksum(check_split[0]);
+        if (checksum.size() == 0) return nullptr; // Invalid characters in payload
```
Done.

MeshCollider commented on src/script/descriptor.cpp:
```diff
@@ -20,6 +20,125 @@
+     *  - As many alphabetic characters are in the same group (while following the above restrictions).
+     *
+     * If p(x) gives the position of a character c in this character set, every group of 3 characters
+     * (a,b,c) is encoded as the 4 symbols (p(a) & 31, p(b) & 31, p(c) & 31, (p(a) / 32) + 3 * (p(b) / 32) + 9 * (p(c) / 32).
```
I'm not sure if I'm reading this wrong, but isn't the fourth symbol in the wrong order, shouldn't this be `(p(c) / 32) + 3 * (p(b) / 32) + 9 * (p(a) / 32)`?

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/15368/commits/fd637be8d21a606e98c037b40b268c4a1fae2244

laanwj commented:
utACK fd637be8d21a606e98c037b40b268c4a1fae2244
